### PR TITLE
[AIR/Data] Revamping APIs for Offline Batch Inference

### PR DESCRIPTION
**Pull request by amogkam**
Reduce the friction for batch inference of pre-trained by updating the APIs to do the following:
1. Remove AIR Checkpoints from the critical path for batch prediction.
2. Simpler interface for chaining multiple preprocessing and prediction stages.

### CODE REVIEW

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:127]:
```suggestion
    .map_predictor(predictor)
```

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:92]: For sake of keeping the scope narrow, I think we should not alter the preprocessor APIs in this REP.

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:135]: This breaks training->serving compatibility, so I think we should punt.

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:97]: The phrasing here is a bit confusing... are you saying that basically users don't need to know about Checkpoints in order to do batch inference with {AIR / custom predictors} in the new proposal? That seems like a win.

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:54]: Maybe it should always be a lambda. I'm not sure if most predictors are copyable.

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:135]: This is just changing the batch inference examples to the new API. The existing training->Ray Serve workflow remains the same?

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md]: removed

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md]: updated

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [reply reps/2023-03-10-batch-inference.md:128]: Got it. Agree batch inference specific workloads can avoid the Preprocessor class entirely.

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:54]:
```suggestion
def map_predictor(create_predictor: Callable[[], Predictor]], 
```

Not sure if we have an existing naming convention for factories.

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:90]: Also remove this as it's needed for Train->Serve?

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:115]: Can you also add an example of how you'd implement a custom predictor? (hopefully it is no harder than using a custom callable cls)

Comment (ericl) [reply reps/2023-03-10-batch-inference.md:115]: I think in general we should be of the mindset that custom predictor/etc. should be presented as first-class in our APIs, and we only have framework-specific helpers like TorchPredictor as shortcuts. This avoids the problem where users do not know how to find the "escape hatch".

File: reps/2023-03-10-batch-inference.md
Comment (gjoliver) [inline reps/2023-03-10-batch-inference.md:119]: like the new dataset-centric way of doing things.
a stupid question, why do we have to differentiate preprocessor vs predictor?
can they all simply use the map_batches() api?
```
predictions = ds.read_parquet("s3://data")
    .map_batches(lambda x: x+1)
    .map_batches(TorchVisionPreprocessor(torchvision.transforms.Crop(224, 224)))
    .map_batches(TorchPredictor(pretrained_resnet()))
```
?

Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:119]: 1. Avoid overloading the existing `map_batches` API and make the API more explicit
2. Not all of the `map_batches` arguments are compatible or needed for preprocessors/predictors.

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md]: good point, removed

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:93]: added

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:114]: added

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:196]: Hmm this is a bit confusing, since you specify "numpy" both here and in the batch format arg. If one is mismatched, I guess you get an error?

Maybe it would be better to have just recommending overriding the predict() method.

Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:196]: oh no, the `batch_format` arg is specified in `map_batches` for the preprocessing logic. `map_predictor` doesn't take in a batch format.

**Comment by richardliaw** [comment]
Usability detail- it may be easy for a user to shoot themselves in a foot by serializing a large model into the scope of the function.

Can we make sure we have a section in the docs + an appropriate error message for this failure mode?

**Comment by richardliaw** [comment]
Would it make sense for map_predictor to take in a predictor instead of a lambda currying a predictor?

**Comment by amogkam** [comment]
Yeah one option can be to allow both:
1. Pass in a Predictor instance. This works for small models, etc. for better usability
2. Pass in a predictor_creator for large models or Predictors created from AIR Checkpoints. 

We can fail if attempting to serialize a very large model via the same mechanism as `tune.with_parameters`

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [reply reps/2023-03-10-batch-inference.md:196]: I guess I'm saying--- in this example (and other examples we create), why not override the plain `predict` class instead? That way, the term "numpy" shows up only once, and the user doesn't need to understand the automatic batch format choosing magic in the Predictor class.

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:12]: We don't show how this is used in the REP--- should we either remove this line, or incorporate it into the REP?

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:196]: Not sure if I'm understanding...we still want to provide both a numpy path and a pandas path right?

When implementing a custom predictor, the user can pick one (or both) to implement. The term numpy is still showing up only in the Predictor. `map_predictor` does not take in a `batch_format` argument.

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md]: removed

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [reply reps/2023-03-10-batch-inference.md:206]: Stepping back, I'm kind of saying the entire way we have designed this is wrong...

Predictor shouldn't query dataset for it's "dataset format". The predictor should say, "this is the batch format I need", and the dataset should internally optimize to minimize format conversions.

In this example, the user ideally should be able to simply write:

```
class MyPrep(Preprocessor):
   batch_format = "numpy"
   
   def predict(self, data):
        assert data is in numpy format
```

Comment (ericl) [reply reps/2023-03-10-batch-inference.md:206]: In other words: why is this so much more complicated than the map_batches() signature? It shouldn't be.

Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:206]: Yes completely agree on what you are saying. But I don't think the API needs to change to achieve this.

> The predictor should say, "this is the batch format I need"

This is achieved by the Predictor either implementing `_predict_pandas` or `_predict_numpy`.

I think the question is should this distinction be in the form of 2 separate methods, or overloading onto a single `predict` method, and having a `batch_format` class attribute.

Personally, I slightly prefer the former just to be more explicit.

Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:206]: Either way, we will be removing the dataset format querying. But this is more of an implementation change than an API change.

Comment (ericl) [reply reps/2023-03-10-batch-inference.md:206]: OK, it seems fine for now if we remove batch format from map_predictor().

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
LGTM. Feel free to submit to the list for a vote.

**Comment by zhe-thoughts** [comment]
High level question: for pre-trained models, what are the main benefits of using `map_preprocessor` and `map_predictor` vs. directly using `map_batches`?

@amogkam @ericl  Thanks

**Comment by amogkam** [comment]
Mainly so that we can provide certain guardrails for users when they use these APIs so they don't shoot themselves in the foot. Also, gives us APIs to add more prediction specific optimizations.

You can see some prior discussion here https://github.com/ray-project/ray/issues/32929#issuecomment-1450573546

**Comment by ericl** [comment]
Re: `map_batches` vs `map_preprocessor` + `map_predictor`, there has been consistent feedback that this is somewhat confusing, because there are now multiple methods that do something very similar. We might be able to achieve the original goal of providing guardrails in other ways.

For _preprocessor_, I'd propose going back to the original idea of implementing `__call__` on Preprocessor classes, and requiring users to call it as follows:

```
ds.map_batches(preprocessor)
```

To choose a good batch format for the preprocessor, Datasets can call `preprocessor.preferred_batch_format()` (this can be a generic interface for specifying batch format on callable) to choose the default batch format. We can also warn at runtime if the user chooses a batch format that isn't preferred.

For _predictor_, we can also implement `__call__` and ensure predictors can be serialized in the common case. For delayed construction of predictor, we can tell users to write a callable class wrapper:

```
# Note: an error should be raised if you don't specify actor pool strategy
ds.map_batches(predictor, compute=ray.data.ActorPoolStrategy(8, 8))

-or-

class MyWrapper:
   def __init__(self):
        self.predictor = MyPredictor(args)

   def __call__(self, batch):
       return self.predictor.predict(batch)

# Note: an error should be raised if you don't specify actor pool strategy
ds.map_batches(MyWrapper, compute=ray.data.ActorPoolStrategy(8, 8))
```

We can also rely on `preferred_batch_format()` for the predictor to choose a good batch format / print a warning to the user if a sub-optimal batch format is used.

Thoughts?

**Comment by amogkam** [comment]
Yeah I'm supportive of this, and goes back to the initial proposal.

I would like to figure out the details of `batch_format` some more though. In your proposal, how would you distinguish between sub-optimal vs. not suboptimal batch format?

Is it just 
```python
if batch_format != preprocessor.determine_transform_to_use():
    raise ValueError(...)
```

**Comment by ericl** [comment]
> I would like to figure out the details of batch_format some more though. In your proposal, how would you distinguish between sub-optimal vs. not suboptimal batch format?

I was thinking that first, Datasets can query `.preferred_batch_format()` on the class passed to it. Within the Preprocessor, in the `__call__` you can check `isinstance(batch, my_preferred_type)`. If it's a mismatch, log once that the batch format is not optimal.

**Comment by amogkam** [comment]
Yeah that sounds good to me. For the wrapped Predictor case, where we don't have direct access to the instance, we would have to directly rely on the runtime check for these warnings.

**Comment by ericl** [comment]
copying from committer list: REP approved, conditional on changes above.

**Review by gjoliver — APPROVED** [review]

**Comment by c21** [comment]
IMO for option 5: this `withColumns()` API semantic is quite complicated… (this would probably confuse people a lot coming from Data background as me). 

I vote for option 1: `TorchPredictor(model, feature_columns=["feature"], keep_columns=["label"])`, and we are doing all the stuff under the hood inside `map_batches`. If no preprocessor defined in checkpoint, we can also push down the feature and keep columns as optimizer rule.

**Comment by ericl** [comment]
I'm in favor of Option 2/3, since I think a little bit of extra boilerplate code is easier to understand and adapt for new users than opaque API args. It's also easy to go from Option 2 to Option 1 at a later point in time if this turns out to be a common pain point, but otherwise harder to go the other direction.

I also don't think Option 3 requires deprecating anything, since we already require users to construct a custom callable if they are delaying the construction of their Predictor. We would just be extending this to tell users to construct a custom callable if they need either delayed construction or this label passthrough.

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:296]: Should this always be true?

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:334]: Is the idea this is called on first use of the predictor? Should it be internal?

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:392]: Can we make this more explicit somehow? Also, we shouldn't expose `setup` publicly right?

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:245]: any suggestions?

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:296]: Idts, for Preprocessors we don't want to use callable classes.

But for Predictors, yes, should always be True

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:331]: It should never need to be called by users.

But, inside `map_batches` it is called here: https://github.com/ray-project/enhancements/pull/25/files#diff-0b39e135189387c4d60ea6f6d574d05c99eeb634ef29e2abae5a92442ce326fdR388

File: reps/2023-03-10-batch-inference.md
Comment (pcmoritz) [inline reps/2023-03-10-batch-inference.md:57]: What is the rationale behind this? As a user I would prefer if I can do
```
predictions = ds.map_batches(predictor)
```
and it does the right thing. Why do I need to care about ActorPoolStrategy?

Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:57]: The default `map_batches` API is TaskPoolStrategy.

If we want to make `ActorPoolStrategy` the default, then we'd have to introduce a new API (like `map_predictor`) which we decided not to do in order to have a single unified API for data processing.

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [reply reps/2023-03-10-batch-inference.md:331]: I don't think we need this though right? For non-lazy Predictors, there is no setup. For lazy predictors, you can lazily do the setup on the first call.

Hence, we can eliminate this API.

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:288]: Shall we have a generic validate_args(kwargs) function instead of adding them one at a time?

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:272]:
```suggestion
    def get_batch_format(self, batch_format: Optional[str]) -> Optional[str]:
```

(It's not just validating)

**Review by ericl — COMMENTED** [review]
I have some blocking comments on BatchMappable. From the process perspective though, it's not critical that we resolve BatchMappable implementation details here--- we could instead say this is to be decided during PR review, and remove the specific detail from this REP.

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [inline reps/2023-03-10-batch-inference.md:122]: I think we want to support a `lazy=True` flag here, right? Otherwise, this wouldn't support deferred loading and/or may be confusing for users if it defaulted to lazy loading.

**Comment by zhe-thoughts** [comment]
@amogkam could you do DCO https://www.secondstate.io/articles/dco/

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:296]: How would that work?

Comment (ericl) [reply reps/2023-03-10-batch-inference.md:296]: We could have a validator like

```
def validate_args(self, **kwargs):
   pass
```

And Datasets could pass in a kwargs dict like `{"compute": ..., "batch_size": ..., ...}`.

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:331]: It would be preferable to call this separately here: https://github.com/ray-project/enhancements/pull/25/files#diff-0b39e135189387c4d60ea6f6d574d05c99eeb634ef29e2abae5a92442ce326fdR388

Otherwise, we can't start the initialization until there are actual prediction tasks to dispatch.

This is purely a developer facing API, so has no user impact.

Comment (ericl) [reply reps/2023-03-10-batch-inference.md:331]: I see, that makes sense (create as soon as actor is created). Would be good to add this as a motivation for this API.

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:122]: This is the old BatchPredictor code path. This doesn't need to change right?

Comment (ericl) [reply reps/2023-03-10-batch-inference.md:122]: Ah, right, I mean to comment on the new path.

File: reps/2023-03-10-batch-inference.md
Comment (ericl) [reply reps/2023-03-10-batch-inference.md:245]: BatchUDF?

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:122]: Got it. I think we should just make lazy the default, so that users don't always need to do `lazy=True` for batch prediction.

Comment (ericl) [reply reps/2023-03-10-batch-inference.md:122]: I'm ok with that. You had a concern earlier that this would make single-process prediction benchmarks misleading though right? Since predict() would trigger loading and that could throw off timing of scripts.

File: reps/2023-03-10-batch-inference.md
Comment (pcmoritz) [reply reps/2023-03-10-batch-inference.md:57]: If a basic batch inference user needs to know about `ActorPoolStrategy` I think we are failing our users, it is not something the user should know about (this is an abstraction leak).

Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:57]: @pcmoritz your recommendation is go with a separate `map_predictor` API?

Comment (ericl) [reply reps/2023-03-10-batch-inference.md:57]: I don't think it's very easy to avoid this abstraction leak *entirely*. Here are a few things we've learned from previous users of this API:
- No one uses the highest level BatchPredictor API, which does try to hide most of underlying details.
- Some users have expressed that they prefer lower-level APIs (such as Ray core), because they understand better what Ray is doing under the hood.

As for creating a special `map_predictor()` method, I am OK with that, but we have gotten strong feedback that it's confusing why there are two very similar map methods. However, upon reflection we may be able to address this concern by explicitly documenting this as the syntactic sugar / simpler form of `map_batches()` for batch inference, similar to how `map` is sugar for `map_batches`.

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:122]: Yeah on second thought, I don't think it's that big of a concern, and will just make the API more confusing.

Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:131]: removed the `lazy` arg and made lazy by default

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md]: Clarified in the docstring for `setup`

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md]: that's already taken by https://github.com/ray-project/ray/blob/master/python/ray/data/block.py#L115 😅

File: reps/2023-03-10-batch-inference.md
Comment (amogkam) [reply reps/2023-03-10-batch-inference.md]: I think it's fine for now since it's just 2 (batch format, compute strategy). If we add more, then we can look into consolidating in the future.

File: reps/2023-03-10-batch-inference.md
Comment (zhe-thoughts) [reply reps/2023-03-10-batch-inference.md:57]: It's a non-trivial tradeoff. For now my preference is to go with `map_batches` with `ActorPoolStrategy`. If we see enough user feedback that a separate `map_predictor` is acceptable, we can do that down the road. I feel it is an incremental path. @pcmoritz what do you think?

Another reason why I prefer this sequence is, for really simple use inference use cases, "it works" without the `ActorPoolStrategy` argument, just slower. @amogkam what is the rough performance difference with vs. without it?

Comment (pcmoritz) [reply reps/2023-03-10-batch-inference.md:57]: Why not use `map_batches` and if the argument is a class (like an actor class or Predictor class), it will use the `ActorPoolStrategy`? I don't see the tradeoff here. It can have a reasonable default like "auto determine the number of actors" and if people want to override, they can specify the `ActorPoolStrategy`.

Comment (ericl) [reply reps/2023-03-10-batch-inference.md:57]: There is the tradeoff between explicitness/understandability and ease of getting started.

If we go with auto determining the number of actors (perhaps with a warning/tip on how to configure it), most users will probably start with this code path.

However if we raise an error and tell the user to specify ActorPoolStrategy(size=N), this starts users down the more explicit and understandable code path, though it adds an additional onboarding step.

The current API is preferring the more explicit route, because (1) users have tended to prefer explicit / easy to understand solutions, and (2) autoscaling introduces performance questionmarks (e.g., slow scale up) and/or complexities (e.g., predictive autoscaling) that can trip up new users trying to get something basic to work.

Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:57]: This REP is just being consistent with the current behavior (https://github.com/ray-project/ray/blob/master/python/ray/data/dataset.py#L342-L352) and is not making a stance on whether there should be a default path for ActorPoolStrategy.

Whether there should be a default path for ActorPoolStrategy is an important discussion to have, but is orthogonal to this REP. Let's continue this discussion as a separate change on #api-changes. If we do decide to have a default ActorPoolStrategy path, then this REP will be modified accordingly.

Comment (pcmoritz) [reply reps/2023-03-10-batch-inference.md:57]: In the `before` code, the user didn't need to specify a number of actors, and now they have to, that looks like a clear regression with this proposal to me.

How was the number of actors determined beforehand?

Comment (amogkam) [reply reps/2023-03-10-batch-inference.md:57]: I'm specifically referring to this proposal that you suggested

> Why not use map_batches and if the argument is a class (like a Predictor class or another class that gets called), it will use the ActorPoolStrategy?

which is a more broader change that goes beyond this REP

**Comment by zhe-thoughts** [comment]
Discussed with @amogkam  @ericl @pcmoritz offline. We will go with the API in the current proposal, and try to further simplify it in the future.

**Comment by zhe-thoughts** [comment]
The further improvement for `ActorPool` is tracked here: https://github.com/ray-project/ray/issues/34034

**Comment by GokuMohandas** [comment]
@amogkam sorry to dig this PR back up but are `map_preprocessor` and `map_predictor` under the 2.4 release (as per [this](https://github.com/ray-project/enhancements/blob/128dc3b907ed61532ea6e9c969e44d3fd9c80001/reps/2023-03-10-batch-inference.md#compatibility-deprecation-and-migration-plan)). If not yet, do you know if/how I can apply a predictor to a dataset for batch inference easily (without creating a separate callable class UDF)?

**Comment by amogkam** [comment]
No they are not in 2.4...currently to use a Predictor with Ray Data directly (not using BatchPredictor), you would do it like so

```python
class Wrapper:
    def __init__(self):
        self.predictor = TorchPredictor(...)

    def __call__(self, batch):
        return self.predictor.predict(batch, ...)


ds.map_batches(Wrapper, ...)
```

**Pull request merged**
